### ISSUE TITLE
Latest google terraform provider versions don't build?

### ISSUE THREAD

**Issue by christian-marie** [NixOS/nixpkgs#95886]
I'm trying to build the latest google terraform provider by overriding the existing versions, here is a shell.nix reproducing the error I'm hitting:

```
{ pkgs ? import <nixpkgs> {} }:
let 
  new_google = pkgs.terraform-providers.google.overrideAttrs (old: rec {
    inherit (old) owner repo;
    name = "${repo}-${version}";
    version = "3.35.0";
    rev = "v3.35.0";
    sha256 = "1qwkckprv708qp5p0dklhpcnawaccllbixjln1sl0dhjk2qr45w4";
    src = pkgs.fetchFromGitHub { inherit owner repo rev sha256; };
  });
in
  pkgs.mkShell {
    buildInputs = [ new_google ];
  }
```
```
$ nix-shell
these derivations will be built:
  /nix/store/a7c6f97bxjqiqp40xw9kmr4b4zg7m2wa-terraform-provider-google-3.35.0.drv
building '/nix/store/a7c6f97bxjqiqp40xw9kmr4b4zg7m2wa-terraform-provider-google-3.35.0.drv'...
unpacking sources
unpacking source archive /nix/store/lfj13cgq9kbv2f1kcmbd6pv9lchmrli9-source
source root is source
patching sources
configuring
building
go: inconsistent vendoring in /private/var/folders/kz/8sv89dl97sx811td5nr_lh28bx37d1/T/nix-build-terraform-provider-google-3.35.0.drv-0/source:
        cloud.google.com/go/bigtable@v1.5.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/apparentlymart/go-cidr@v1.0.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/client9/misspell@v0.3.4: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/davecgh/go-spew@v1.1.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/dnaeon/go-vcr@v1.0.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/dustinkirkland/golang-petname@v0.0.0-20170921220637-d3c2ba80e75e: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/gammazero/deque@v0.0.0-20180920172122-f6adf94963e4: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/gammazero/workerpool@v0.0.0-20181230203049-86a96b5d5d92: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/gogo/protobuf@v1.2.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/golangci/golangci-lint@v1.18.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/hashicorp/errwrap@v1.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/hashicorp/go-cleanhttp@v0.5.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/hashicorp/go-multierror@v1.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/hashicorp/go-version@v1.2.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/hashicorp/terraform-plugin-sdk@v1.11.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/mitchellh/hashstructure@v1.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/sirupsen/logrus@v1.2.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/stoewer/go-strcase@v1.0.2: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        github.com/terraform-providers/terraform-provider-random@v0.0.0-20190925200408-30dac3233094: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/net@v0.0.0-20200625001655-4c5254603344: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        golang.org/x/oauth2@v0.0.0-20200107190931-bf48bf16ab8d: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
        google.golang.org/api@v0.29.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt

run 'go mod vendor' to sync, or use -mod=mod or -mod=readonly to ignore the vendor directory
builder for '/nix/store/a7c6f97bxjqiqp40xw9kmr4b4zg7m2wa-terraform-provider-google-3.35.0.drv' failed with exit code 1
error: build of '/nix/store/a7c6f97bxjqiqp40xw9kmr4b4zg7m2wa-terraform-provider-google-3.35.0.drv' failed
```

Is this possibly because the vendor/ directory was removed? https://github.com/hashicorp/terraform-provider-google/commit/8bf122fb6996859297a563fc4c95ceed0c85a307

I noticed this override: https://github.com/NixOS/nixpkgs/blob/3cece2b33b43ad2766b25053ed43d8fc8eca914a/pkgs/applications/networking/cluster/terraform-providers/default.nix#L23

I'm looking for any way to build a more recent google terraform provider. Ideas?

**Comment by christian-marie** [comment]
Hi, to rephrase this as not a question - please may we have some more recent terraform providers?

The current ones are quite old. I'm happy to do the work to update them but would appreciate some guidance around how these modules are meant to work - perhaps @zimbatm  knows if this is an expected explosion due to the upstream vendor/ change? I'm not particularly familiar with go modules.

**Comment by JamesReynolds** [comment]
I found a solution to this, the problem occurs because "${name}" is used within a string in "go-modules" and "${go-modules}" is used within a string in "configurePhase".

I added something like this to my overlay (I'm using minio):

```
    go-modules = super.stdenv.mkDerivation (let modArgs = {
          name = "minio-${version}-go-modules";
          nativeBuildInputs = [ super.go super.git super.cacert ];
          inherit src;
          patches = [];
          preBuild = "";
          sourceRoot = "";
          GO111MODULE = "on";
          impureEnvVars = super.lib.fetchers.proxyImpureEnvVars ++ [
            "GIT_PROXY_COMMAND" "SOCKS_SERVER"
          ];
          configurePhase = ''
            runHook preConfigure

            export GOCACHE=$TMPDIR/go-cache
            export GOPATH="$TMPDIR/go"
            cd "./"
            runHook postConfigure
          '';
          buildPhase = ''
            runHook preBuild
            go mod vendor
            mkdir -p vendor
            runHook postBuild
          '';
          installPhase = ''
            runHook preInstall
            # remove cached lookup results and tiles
            cp -r --reflink=auto vendor $out
            runHook postInstall
          '';
          dontFixup = true;
        }; in modArgs // (
            {
              outputHashMode = "recursive";
              outputHashAlgo = "sha256";
              outputHash = "1d8wd5p6005v0cvq88124hvz15raha9gkym0pwymcjh6wr9k7x46";
            }
        ));
        configurePhase = ''
          runHook preConfigure
          export GOCACHE=$TMPDIR/go-cache
          export GOPATH="$TMPDIR/go"
          export GOSUMDB=off
          export GOPROXY=off
          cd "$modRoot"
          rm -rf vendor
          ln -s ${go-modules} vendor
          runHook postConfigure
        '';
```

Its far from ideal - unfortunately I can't find a way to hook and pull out the modules to push into the configure phase without copying all of the code to build the modules. It works... but its not great.
